File: game/obstacle.py
"""
Obstacle Entities and Spawner System for Terminal Runner
Features 5 distinct obstacle types (Ground Spikes, Security Drones, Data Pylons,
Laser Barriers, and Glitch Nodes) with procedural cyberpunk graphics.
"""

import math
import random
import pygame
from game.config import (
    SCREEN_WIDTH, GROUND_Y, COLOR_NEON_RED, COLOR_NEON_AMBER,
    COLOR_NEON_PINK, COLOR_NEON_CYAN, COLOR_NEON_PURPLE, COLOR_NEON_GREEN,
    COLOR_WHITE, COLOR_SURFACE, COLOR_BG
)


class BaseObstacle:
    """Base class for all obstacles."""

    def __init__(self, x: float, y: float, width: int, height: int, obs_type: str):
        self.x = float(x)
        self.y = float(y)
        self.width = width
        self.height = height
        self.obs_type = obs_type
        self.rect = pygame.Rect(int(self.x), int(self.y), width, height)
        self.passed = False
        self.anim_timer = 0.0

    def update(self, dt: float, world_speed: float):
        """Moves obstacle to the left based on world speed."""
        self.x -= world_speed * dt * 60.0
        self.anim_timer += dt
        self._update_rect()

    def _update_rect(self):
        self.rect.x = int(self.x)
        self.rect.y = int(self.y)

    def draw(self, surface: pygame.Surface):
        raise NotImplementedError


class FirewallSpikes(BaseObstacle):
    """Ground-based neon laser spikes (Single, Double, or Triple cluster)."""

    def __init__(self, x: float, cluster_size: int = 1):
        self.cluster_size = cluster_size
        width = 24 * cluster_size
        height = random.choice([36, 42])
        y = GROUND_Y - height
        super().__init__(x, y, width, height, "spikes")
        # Fair collision rect margin
        self.hitbox_margin_x = 3
        self.hitbox_margin_y = 2

    def _update_rect(self):
        self.rect = pygame.Rect(
            int(self.x + self.hitbox_margin_x),
            int(self.y + self.hitbox_margin_y),
            self.width - (self.hitbox_margin_x * 2),
            self.height - self.hitbox_margin_y
        )

    def draw(self, surface: pygame.Surface):
        surf = pygame.Surface((self.width + 4, self.height + 4), pygame.SRCALPHA)
        spike_w = 24

        for i in range(self.cluster_size):
            sx = i * spike_w
            # Spike triangle points
            top_pt = (sx + 12, 2)
            left_pt = (sx + 2, self.height)
            right_pt = (sx + 22, self.height)

            # Animated glowing pulse
            glow_intensity = int(180 + 75 * math.sin(self.anim_timer * 8 + i))
            color_core = (255, glow_intensity, 0)

            # Dark inner backing
            pygame.draw.polygon(surf, COLOR_SURFACE, [top_pt, left_pt, right_pt])
            # Outer neon frame
            pygame.draw.polygon(surf, COLOR_NEON_AMBER, [top_pt, left_pt, right_pt], width=2)
            # Center energy beam
            pygame.draw.line(surf, color_core, (sx + 12, self.height - 2), (sx + 12, 8), 2)
            # Tip glow
            pygame.draw.circle(surf, COLOR_WHITE, (sx + 12, 4), 2)

        surface.blit(surf, (int(self.x), int(self.y)))


class SecurityDrone(BaseObstacle):
    """Air-based patrol drone with scanner cone and hovering dynamics."""

    def __init__(self, x: float, altitude_type: str = "mid"):
        # Altitude types: "low" (must duck or precise jump), "high" (can run/duck under, jump will hit)
        self.altitude_type = altitude_type
        width = 44
        height = 30

        if altitude_type == "low":
            base_y = GROUND_Y - 58  # Head height (Must duck!)
        else:
            base_y = GROUND_Y - 96  # High jump zone

        super().__init__(x, base_y, width, height, "drone")
        self.base_y = base_y
        self.hover_offset = 0.0

    def update(self, dt: float, world_speed: float):
        super().update(dt, world_speed)
        # Gentle floating hover motion
        self.hover_offset = math.sin(self.anim_timer * 6.0) * 4.0
        self.y = self.base_y + self.hover_offset
        self._update_rect()

    def _update_rect(self):
        margin = 4
        self.rect = pygame.Rect(
            int(self.x + margin),
            int(self.y + margin),
            self.width - (margin * 2),
            self.height - (margin * 2)
        )

    def draw(self, surface: pygame.Surface):
        surf = pygame.Surface((self.width + 10, self.height + 30), pygame.SRCALPHA)
        center_x = self.width // 2 + 5
        center_y = 15

        # Downward scanning cone
        scanner_alpha = int(40 + 25 * math.sin(self.anim_timer * 10))
        scan_pts = [(center_x, center_y + 8),
                    (center_x - 18, self.height + 25),
                    (center_x + 18, self.height + 25)]
        pygame.draw.polygon(surf, (*COLOR_NEON_RED, scanner_alpha), scan_pts)

        # Drone Main Body
        drone_rect = pygame.Rect(center_x - 16, center_y - 8, 32, 16)
        pygame.draw.rect(surf, COLOR_SURFACE, drone_rect, border_radius=4)
        pygame.draw.rect(surf, COLOR_NEON_PINK, drone_rect, width=2, border_radius=4)

        # Glowing Red Sensor Eye
        eye_color = COLOR_WHITE if int(self.anim_timer * 8) % 2 == 0 else COLOR_NEON_RED
        pygame.draw.circle(surf, eye_color, (center_x, center_y), 4)

        # Micro thrusters on sides
        rotor_phase = self.anim_timer * 20.0
        rotor_w = max(2, int(abs(math.sin(rotor_phase)) * 10))
        # Left rotor
        pygame.draw.line(surf, COLOR_NEON_CYAN, (center_x - 20 - rotor_w, center_y - 6),
                         (center_x - 20 + rotor_w, center_y - 6), 2)
        # Right rotor
        pygame.draw.line(surf, COLOR_NEON_CYAN, (center_x + 20 - rotor_w, center_y - 6),
                         (center_x + 20 + rotor_w, center_y - 6), 2)

        surface.blit(surf, (int(self.x - 5), int(self.y)))


class DataPylon(BaseObstacle):
    """Tall vertical server rack barrier with sparking high-voltage nodes."""

    def __init__(self, x: float):
        width = 30
        height = 62
        y = GROUND_Y - height
        super().__init__(x, y, width, height, "pylon")

    def _update_rect(self):
        margin_x = 3
        margin_y = 2
        self.rect = pygame.Rect(
            int(self.x + margin_x),
            int(self.y + margin_y),
            self.width - (margin_x * 2),
            self.height - margin_y
        )

    def draw(self, surface: pygame.Surface):
        surf = pygame.Surface((self.width + 4, self.height + 4), pygame.SRCALPHA)
        # Pillar structure
        pylon_rect = pygame.Rect(2, 2, self.width, self.height)
        pygame.draw.rect(surf, COLOR_SURFACE, pylon_rect, border_radius=2)
        pygame.draw.rect(surf, COLOR_NEON_PURPLE, pylon_rect, width=2, border_radius=2)

        # Server drive slots with blinking LEDs
        for slot in range(4):
            slot_y = 8 + slot * 12
            pygame.draw.line(surf, (45, 55, 72), (6, slot_y), (self.width - 2, slot_y), 2)
            led_color = COLOR_NEON_GREEN if (int(self.anim_timer * 6) + slot) % 3 == 0 else COLOR_NEON_CYAN
            pygame.draw.rect(surf, led_color, (self.width - 6, slot_y - 1, 3, 3))

        # Top sparking electrode
        pygame.draw.circle(surf, COLOR_WHITE, (self.width // 2 + 2, 4), 3)
        if int(self.anim_timer * 12) % 2 == 0:
            pygame.draw.line(surf, COLOR_NEON_CYAN, (self.width // 2 - 4, 2), (self.width // 2 + 8, 2), 1)

        surface.blit(surf, (int(self.x), int(self.y)))


class LaserBarrier(BaseObstacle):
    """Overhead electric laser beam (Forces the player to duck/slide)."""

    def __init__(self, x: float):
        width = 70
        height = 36
        y = GROUND_Y - 60  # Suspended overhead
        super().__init__(x, y, width, height, "laser")

    def _update_rect(self):
        # Precise collision zone
        margin_x = 2
        margin_y = 4
        self.rect = pygame.Rect(
            int(self.x + margin_x),
            int(self.y + margin_y),
            self.width - (margin_x * 2),
            self.height - (margin_y * 2)
        )

    def draw(self, surface: pygame.Surface):
        surf = pygame.Surface((self.width + 6, self.height + 10), pygame.SRCALPHA)

        # Left & right emitter nodes
        pygame.draw.rect(surf, COLOR_SURFACE, (2, 4, 10, 24), border_radius=2)
        pygame.draw.rect(surf, COLOR_NEON_RED, (2, 4, 10, 24), width=1, border_radius=2)

        pygame.draw.rect(surf, COLOR_SURFACE, (self.width - 8, 4, 10, 24), border_radius=2)
        pygame.draw.rect(surf, COLOR_NEON_RED, (self.width - 8, 4, 10, 24), width=1, border_radius=2)

        # Glowing Laser Beams
        beam_y1 = 12
        beam_y2 = 20
        # Laser core
        pygame.draw.line(surf, COLOR_NEON_RED, (12, beam_y1), (self.width - 8, beam_y1), 4)
        pygame.draw.line(surf, COLOR_WHITE, (12, beam_y1), (self.width - 8, beam_y1), 1)

        pygame.draw.line(surf, COLOR_NEON_RED, (12, beam_y2), (self.width - 8, beam_y2), 4)
        pygame.draw.line(surf, COLOR_WHITE, (12, beam_y2), (self.width - 8, beam_y2), 1)

        # Electricity crackle between beams
        if int(self.anim_timer * 15) % 2 == 0:
            mid_x = random.randint(16, self.width - 16)
            pygame.draw.line(surf, COLOR_NEON_CYAN, (mid_x, beam_y1), (mid_x + random.randint(-4, 4), beam_y2), 2)

        surface.blit(surf, (int(self.x), int(self.y)))


class GlitchNode(BaseObstacle):
    """Digital glitch block with animated shifting wireframe geometry."""

    def __init__(self, x: float):
        width = 36
        height = 36
        y = GROUND_Y - height
        super().__init__(x, y, width, height, "glitch")

    def _update_rect(self):
        self.rect = pygame.Rect(int(self.x + 3), int(self.y + 3), self.width - 6, self.height - 6)

    def draw(self, surface: pygame.Surface):
        surf = pygame.Surface((self.width + 4, self.height + 4), pygame.SRCALPHA)
        # Shifting glitch coordinates
        offset_x = random.choice([-2, 0, 2]) if random.random() < 0.25 else 0
        offset_y = random.choice([-1, 0, 1]) if random.random() < 0.25 else 0

        rect = pygame.Rect(2 + offset_x, 2 + offset_y, self.width, self.height)
        pygame.draw.rect(surf, (20, 10, 30), rect)
        pygame.draw.rect(surf, COLOR_NEON_PINK, rect, width=2)
        # Inner cybernetic cross
        pygame.draw.line(surf, COLOR_NEON_CYAN, (rect.left, rect.centery), (rect.right, rect.centery), 1)
        pygame.draw.line(surf, COLOR_NEON_CYAN, (rect.centerx, rect.top), (rect.centerx, rect.bottom), 1)

        surface.blit(surf, (int(self.x), int(self.y)))


class ObstacleSpawner:
    """Manages adaptive obstacle generation, spacing, and progression."""

    def __init__(self):
        self.obstacles = []
        self.spawn_distance_tracker = 0.0
        self.next_spawn_distance = 450.0

    def clear(self):
        self.obstacles.clear()
        self.spawn_distance_tracker = 0.0
        self.next_spawn_distance = 450.0

    def update(self, dt: float, world_speed: float, total_distance: float):
        """Updates obstacles and spawns new ones dynamically."""
        # Update existing obstacles
        for obs in self.obstacles:
            obs.update(dt, world_speed)

        # Remove off-screen obstacles
        self.obstacles = [obs for obs in self.obstacles if obs.x + obs.width > -50]

        # Calculate spawn timing
        self.spawn_distance_tracker += world_speed * dt * 60.0

        if self.spawn_distance_tracker >= self.next_spawn_distance:
            self._spawn_obstacle(total_distance, world_speed)
            self.spawn_distance_tracker = 0.0
            # Adapt next spawn distance based on current speed
            # Minimum distance scales up as speed increases so reaction time remains fair
            min_dist = max(300, int(220 + world_speed * 18))
            max_dist = min_dist + random.randint(120, 280)
            self.next_spawn_distance = random.uniform(min_dist, max_dist)

    def _spawn_obstacle(self, total_distance: float, world_speed: float):
        """Intelligently picks next obstacle type based on difficulty progression."""
        spawn_x = SCREEN_WIDTH + 40

        # Progressive unlock thresholds:
        # Distance < 800: Only single/double spikes and glitch nodes
        # Distance 800 - 2000: Introduce Security Drones (low/high)
        # Distance 2000 - 3500: Introduce Data Pylons & Laser Barriers
        # Distance > 3500: Full hazard suite with triple spikes

        available_types = ["spikes_1", "glitch"]

        if total_distance >= 600:
            available_types.extend(["spikes_2", "drone_high"])
        if total_distance >= 1500:
            available_types.extend(["drone_low", "pylon"])
        if total_distance >= 2800:
            available_types.extend(["laser", "spikes_3"])

        choice = random.choice(available_types)

        if choice == "spikes_1":
            obs = FirewallSpikes(spawn_x, cluster_size=1)
        elif choice == "spikes_2":
            obs = FirewallSpikes(spawn_x, cluster_size=2)
        elif choice == "spikes_3":
            obs = FirewallSpikes(spawn_x, cluster_size=3)
        elif choice == "drone_high":
            obs = SecurityDrone(spawn_x, altitude_type="high")
        elif choice == "drone_low":
            obs = SecurityDrone(spawn_x, altitude_type="low")
        elif choice == "pylon":
            obs = DataPylon(spawn_x)
        elif choice == "laser":
            obs = LaserBarrier(spawn_x)
        else:
            obs = GlitchNode(spawn_x)

        self.obstacles.append(obs)

    def check_collision(self, player_rect: pygame.Rect) -> BaseObstacle:
        """Returns the colliding obstacle if any."""
        for obs in self.obstacles:
            if obs.rect.colliderect(player_rect):
                return obs
        return None

    def draw(self, surface: pygame.Surface):
        """Draws all active obstacles."""
        for obs in self.obstacles:
            obs.draw(surface)

File: src/obstacle.ts
/**
 * Obstacle Entities and Spawner System for Terminal Runner (HTML5 Canvas)
 * Features 5 distinct obstacle types:
 * 1. Firewall Spikes (Single, Double, Triple neon laser ground clusters)
 * 2. Security Drones (Low duck-under / High jump-under with animated scanner cones)
 * 3. Data Pylons (High-voltage server racks with blinking LEDs)
 * 4. Laser Barriers (Overhead dual beams forcing ducking slides)
 * 5. Glitch Nodes (Shifting digital hazard blocks)
 */

import {
  SCREEN_WIDTH, GROUND_Y, COLOR_NEON_RED, COLOR_NEON_AMBER,
  COLOR_NEON_PINK, COLOR_NEON_CYAN, COLOR_NEON_PURPLE, COLOR_NEON_GREEN,
  COLOR_WHITE, COLOR_SURFACE
} from './config.ts';
import { Rect, ObstacleType } from './types.ts';

export abstract class BaseObstacle {
  public x: number;
  public y: number;
  public width: number;
  public height: number;
  public obsType: ObstacleType;
  public rect: Rect;
  public passed: boolean = false;
  public animTimer: number = 0.0;

  constructor(x: number, y: number, width: number, height: number, obsType: ObstacleType) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.obsType = obsType;
    this.rect = { x: Math.floor(x), y: Math.floor(y), width, height };
  }

  public update(dt: number, worldSpeed: number): void {
    this.x -= worldSpeed * dt * 60.0;
    this.animTimer += dt;
    this.updateRect();
  }

  protected updateRect(): void {
    this.rect.x = Math.floor(this.x);
    this.rect.y = Math.floor(this.y);
  }

  public abstract draw(ctx: CanvasRenderingContext2D): void;
}

export class FirewallSpikes extends BaseObstacle {
  public clusterSize: number;
  private marginX = 3;
  private marginY = 2;

  constructor(x: number, clusterSize: number = 1) {
    const width = 24 * clusterSize;
    const height = Math.random() < 0.5 ? 36 : 42;
    const y = GROUND_Y - height;
    super(x, y, width, height, "spikes");
    this.clusterSize = clusterSize;
    this.updateRect();
  }

  protected override updateRect(): void {
    this.rect = {
      x: Math.floor(this.x + this.marginX),
      y: Math.floor(this.y + this.marginY),
      width: this.width - this.marginX * 2,
      height: this.height - this.marginY
    };
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    const spikeW = 24;

    for (let i = 0; i < this.clusterSize; i++) {
      const sx = this.x + i * spikeW;
      const topPt = { x: sx + 12, y: this.y + 2 };
      const leftPt = { x: sx + 2, y: this.y + this.height };
      const rightPt = { x: sx + 22, y: this.y + this.height };

      const glowIntensity = Math.floor(180 + 75 * Math.sin(this.animTimer * 8 + i));
      const colorCore = `rgb(255, ${glowIntensity}, 0)`;

      // Dark inner backing
      ctx.fillStyle = COLOR_SURFACE;
      ctx.beginPath();
      ctx.moveTo(topPt.x, topPt.y);
      ctx.lineTo(leftPt.x, leftPt.y);
      ctx.lineTo(rightPt.x, rightPt.y);
      ctx.closePath();
      ctx.fill();

      // Outer neon frame
      ctx.strokeStyle = COLOR_NEON_AMBER;
      ctx.lineWidth = 2;
      ctx.stroke();

      // Center energy beam
      ctx.strokeStyle = colorCore;
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.moveTo(sx + 12, this.y + this.height - 2);
      ctx.lineTo(sx + 12, this.y + 8);
      ctx.stroke();

      // Tip glow
      ctx.fillStyle = COLOR_WHITE;
      ctx.beginPath();
      ctx.arc(sx + 12, this.y + 4, 2, 0, Math.PI * 2);
      ctx.fill();
    }
    ctx.restore();
  }
}

export class SecurityDrone extends BaseObstacle {
  public altitudeType: 'low' | 'high';
  private baseY: number;
  private margin = 4;

  constructor(x: number, altitudeType: 'low' | 'high' = 'mid' as unknown as 'low') {
    const width = 44;
    const height = 30;
    const baseAltitude = altitudeType === 'low' ? 'low' : 'high';
    const baseY = baseAltitude === 'low' ? GROUND_Y - 58 : GROUND_Y - 96;

    super(x, baseY, width, height, "drone");
    this.altitudeType = baseAltitude;
    this.baseY = baseY;
    this.updateRect();
  }

  public override update(dt: number, worldSpeed: number): void {
    super.update(dt, worldSpeed);
    const hoverOffset = Math.sin(this.animTimer * 6.0) * 4.0;
    this.y = this.baseY + hoverOffset;
    this.updateRect();
  }

  protected override updateRect(): void {
    this.rect = {
      x: Math.floor(this.x + this.margin),
      y: Math.floor(this.y + this.margin),
      width: this.width - this.margin * 2,
      height: this.height - this.margin * 2
    };
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    const centerX = Math.floor(this.x + this.width / 2);
    const centerY = Math.floor(this.y + 15);

    // Downward scanning cone
    const scannerAlpha = (40 + 25 * Math.sin(this.animTimer * 10)) / 255.0;
    ctx.fillStyle = `rgba(255, 34, 68, ${scannerAlpha.toFixed(3)})`;
    ctx.beginPath();
    ctx.moveTo(centerX, centerY + 8);
    ctx.lineTo(centerX - 18, this.y + this.height + 25);
    ctx.lineTo(centerX + 18, this.y + this.height + 25);
    ctx.closePath();
    ctx.fill();

    // Drone Main Body
    ctx.fillStyle = COLOR_SURFACE;
    ctx.beginPath();
    ctx.roundRect(centerX - 16, centerY - 8, 32, 16, 4);
    ctx.fill();

    ctx.strokeStyle = COLOR_NEON_PINK;
    ctx.lineWidth = 2;
    ctx.stroke();

    // Glowing Red Sensor Eye
    const eyeBlink = Math.floor(this.animTimer * 8) % 2 === 0;
    ctx.fillStyle = eyeBlink ? COLOR_WHITE : COLOR_NEON_RED;
    ctx.beginPath();
    ctx.arc(centerX, centerY, 4, 0, Math.PI * 2);
    ctx.fill();

    // Micro thrusters on sides
    const rotorPhase = this.animTimer * 20.0;
    const rotorW = Math.max(2, Math.floor(Math.abs(Math.sin(rotorPhase)) * 10));

    ctx.strokeStyle = COLOR_NEON_CYAN;
    ctx.lineWidth = 2;
    // Left rotor
    ctx.beginPath();
    ctx.moveTo(centerX - 20 - rotorW, centerY - 6);
    ctx.lineTo(centerX - 20 + rotorW, centerY - 6);
    ctx.stroke();

    // Right rotor
    ctx.beginPath();
    ctx.moveTo(centerX + 20 - rotorW, centerY - 6);
    ctx.lineTo(centerX + 20 + rotorW, centerY - 6);
    ctx.stroke();

    ctx.restore();
  }
}

export class DataPylon extends BaseObstacle {
  private marginX = 3;
  private marginY = 2;

  constructor(x: number) {
    const width = 30;
    const height = 62;
    const y = GROUND_Y - height;
    super(x, y, width, height, "pylon");
    this.updateRect();
  }

  protected override updateRect(): void {
    this.rect = {
      x: Math.floor(this.x + this.marginX),
      y: Math.floor(this.y + this.marginY),
      width: this.width - this.marginX * 2,
      height: this.height - this.marginY
    };
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    // Pillar structure
    ctx.fillStyle = COLOR_SURFACE;
    ctx.beginPath();
    ctx.roundRect(this.x, this.y, this.width, this.height, 2);
    ctx.fill();

    ctx.strokeStyle = COLOR_NEON_PURPLE;
    ctx.lineWidth = 2;
    ctx.stroke();

    // Server drive slots with blinking LEDs
    for (let slot = 0; slot < 4; slot++) {
      const slotY = this.y + 8 + slot * 12;
      ctx.strokeStyle = "#2d3748";
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.moveTo(this.x + 4, slotY);
      ctx.lineTo(this.x + this.width - 4, slotY);
      ctx.stroke();

      const ledColor = (Math.floor(this.animTimer * 6) + slot) % 3 === 0 ? COLOR_NEON_GREEN : COLOR_NEON_CYAN;
      ctx.fillStyle = ledColor;
      ctx.fillRect(this.x + this.width - 7, slotY - 2, 3, 3);
    }

    // Top electrode
    ctx.fillStyle = COLOR_WHITE;
    ctx.beginPath();
    ctx.arc(this.x + this.width / 2, this.y + 4, 3, 0, Math.PI * 2);
    ctx.fill();

    if (Math.floor(this.animTimer * 12) % 2 === 0) {
      ctx.strokeStyle = COLOR_NEON_CYAN;
      ctx.lineWidth = 1;
      ctx.beginPath();
      ctx.moveTo(this.x + this.width / 2 - 4, this.y + 2);
      ctx.lineTo(this.x + this.width / 2 + 8, this.y + 2);
      ctx.stroke();
    }

    ctx.restore();
  }
}

export class LaserBarrier extends BaseObstacle {
  private marginX = 2;
  private marginY = 4;

  constructor(x: number) {
    const width = 70;
    const height = 36;
    const y = GROUND_Y - 60; // Suspended overhead (duck/slide under)
    super(x, y, width, height, "laser");
    this.updateRect();
  }

  protected override updateRect(): void {
    this.rect = {
      x: Math.floor(this.x + this.marginX),
      y: Math.floor(this.y + this.marginY),
      width: this.width - this.marginX * 2,
      height: this.height - this.marginY * 2
    };
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    // Left & right emitter nodes
    ctx.fillStyle = COLOR_SURFACE;
    ctx.fillRect(this.x, this.y + 4, 10, 24);
    ctx.strokeStyle = COLOR_NEON_RED;
    ctx.lineWidth = 1;
    ctx.strokeRect(this.x, this.y + 4, 10, 24);

    ctx.fillRect(this.x + this.width - 10, this.y + 4, 10, 24);
    ctx.strokeRect(this.x + this.width - 10, this.y + 4, 10, 24);

    // Glowing Laser Beams
    const beamY1 = this.y + 12;
    const beamY2 = this.y + 20;

    // Laser 1
    ctx.strokeStyle = COLOR_NEON_RED;
    ctx.lineWidth = 4;
    ctx.beginPath();
    ctx.moveTo(this.x + 10, beamY1);
    ctx.lineTo(this.x + this.width - 10, beamY1);
    ctx.stroke();

    ctx.strokeStyle = COLOR_WHITE;
    ctx.lineWidth = 1;
    ctx.beginPath();
    ctx.moveTo(this.x + 10, beamY1);
    ctx.lineTo(this.x + this.width - 10, beamY1);
    ctx.stroke();

    // Laser 2
    ctx.strokeStyle = COLOR_NEON_RED;
    ctx.lineWidth = 4;
    ctx.beginPath();
    ctx.moveTo(this.x + 10, beamY2);
    ctx.lineTo(this.x + this.width - 10, beamY2);
    ctx.stroke();

    ctx.strokeStyle = COLOR_WHITE;
    ctx.lineWidth = 1;
    ctx.beginPath();
    ctx.moveTo(this.x + 10, beamY2);
    ctx.lineTo(this.x + this.width - 10, beamY2);
    ctx.stroke();

    // Electricity crackle
    if (Math.floor(this.animTimer * 15) % 2 === 0) {
      const midX = this.x + 16 + Math.random() * (this.width - 32);
      ctx.strokeStyle = COLOR_NEON_CYAN;
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.moveTo(midX, beamY1);
      ctx.lineTo(midX + (Math.random() * 8 - 4), beamY2);
      ctx.stroke();
    }

    ctx.restore();
  }
}

export class GlitchNode extends BaseObstacle {
  constructor(x: number) {
    const width = 36;
    const height = 36;
    const y = GROUND_Y - height;
    super(x, y, width, height, "glitch");
    this.updateRect();
  }

  protected override updateRect(): void {
    this.rect = {
      x: Math.floor(this.x + 3),
      y: Math.floor(this.y + 3),
      width: this.width - 6,
      height: this.height - 6
    };
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    const offsetX = Math.random() < 0.25 ? (Math.random() < 0.5 ? -2 : 2) : 0;
    const offsetY = Math.random() < 0.25 ? (Math.random() < 0.5 ? -1 : 1) : 0;

    const rx = this.x + offsetX;
    const ry = this.y + offsetY;

    ctx.fillStyle = "#140a1e";
    ctx.fillRect(rx, ry, this.width, this.height);

    ctx.strokeStyle = COLOR_NEON_PINK;
    ctx.lineWidth = 2;
    ctx.strokeRect(rx, ry, this.width, this.height);

    // Inner cross
    ctx.strokeStyle = COLOR_NEON_CYAN;
    ctx.lineWidth = 1;
    ctx.beginPath();
    ctx.moveTo(rx, ry + this.height / 2);
    ctx.lineTo(rx + this.width, ry + this.height / 2);
    ctx.moveTo(rx + this.width / 2, ry);
    ctx.lineTo(rx + this.width / 2, ry + this.height);
    ctx.stroke();

    ctx.restore();
  }
}

export class ObstacleSpawner {
  public obstacles: BaseObstacle[] = [];
  public spawnDistanceTracker: number = 0.0;
  public nextSpawnDistance: number = 450.0;

  public clear(): void {
    this.obstacles = [];
    this.spawnDistanceTracker = 0.0;
    this.nextSpawnDistance = 450.0;
  }

  public update(dt: number, worldSpeed: number, totalDistance: number): void {
    for (const obs of this.obstacles) {
      obs.update(dt, worldSpeed);
    }

    // Remove off-screen obstacles
    this.obstacles = this.obstacles.filter(obs => obs.x + obs.width > -50);

    // Calculate spawn timing
    this.spawnDistanceTracker += worldSpeed * dt * 60.0;

    if (this.spawnDistanceTracker >= this.nextSpawnDistance) {
      this.spawnObstacle(totalDistance, worldSpeed);
      this.spawnDistanceTracker = 0.0;
      const minDist = Math.max(300, Math.floor(220 + worldSpeed * 18));
      const maxDist = minDist + 120 + Math.floor(Math.random() * 160);
      this.nextSpawnDistance = minDist + Math.random() * (maxDist - minDist);
    }
  }

  public spawnObstacle(totalDistance: number, _worldSpeed: number): void {
    const spawnX = SCREEN_WIDTH + 40;
    const availableTypes: string[] = ["spikes_1", "glitch"];

    if (totalDistance >= 600) {
      availableTypes.push("spikes_2", "drone_high");
    }
    if (totalDistance >= 1500) {
      availableTypes.push("drone_low", "pylon");
    }
    if (totalDistance >= 2800) {
      availableTypes.push("laser", "spikes_3");
    }

    const choice = availableTypes[Math.floor(Math.random() * availableTypes.length)];
    let obs: BaseObstacle;

    if (choice === "spikes_1") {
      obs = new FirewallSpikes(spawnX, 1);
    } else if (choice === "spikes_2") {
      obs = new FirewallSpikes(spawnX, 2);
    } else if (choice === "spikes_3") {
      obs = new FirewallSpikes(spawnX, 3);
    } else if (choice === "drone_high") {
      obs = new SecurityDrone(spawnX, "high");
    } else if (choice === "drone_low") {
      obs = new SecurityDrone(spawnX, "low");
    } else if (choice === "pylon") {
      obs = new DataPylon(spawnX);
    } else if (choice === "laser") {
      obs = new LaserBarrier(spawnX);
    } else {
      obs = new GlitchNode(spawnX);
    }

    this.obstacles.push(obs);
  }

  public checkCollision(playerRect: Rect): BaseObstacle | null {
    for (const obs of this.obstacles) {
      if (
        playerRect.x < obs.rect.x + obs.rect.width &&
        playerRect.x + playerRect.width > obs.rect.x &&
        playerRect.y < obs.rect.y + obs.rect.height &&
        playerRect.y + playerRect.height > obs.rect.y
      ) {
        return obs;
      }
    }
    return null;
  }

  public draw(ctx: CanvasRenderingContext2D): void {
    for (const obs of this.obstacles) {
      obs.draw(ctx);
    }
  }
}
